cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/AbstractSchedulerService.java

```diff
@@ -90,280 +94,150 @@ protected final void stopScheduler() {
-      checkInitialized();
-      try {
-        Trigger trigger = scheduler.getTrigger(new TriggerKey(getScheduleId(program, programType, scheduleName)));
-        Preconditions.checkNotNull(trigger);
-
-        scheduler.unscheduleJob(trigger.getKey());
-
-        JobKey jobKey = trigger.getJobKey();
-        if (scheduler.getTriggersOfJob(jobKey).isEmpty()) {
-          scheduler.deleteJob(jobKey);
-        }
-      } catch (SchedulerException e) {
-        throw Throwables.propagate(e);
-      }
-    }
-
-    @Override
-    public void deleteSchedules(Id.Program program, SchedulableProgramType programType) {
-      checkInitialized();
-      try {
-        scheduler.deleteJob(getJobKey(program, programType));
-      } catch (SchedulerException e) {
-        throw Throwables.propagate(e);
-      }
+    ScheduleType type = getScheduleType(program, programType, scheduleName);
+    if (type == null) {
+      throw new UnsupportedOperationException("Schedule could not be found in application specification");
     }
-
-    @Override
-    public ScheduleState scheduleState(Id.Program program, SchedulableProgramType programType, String scheduleName) {
-      checkInitialized();
-      try {
-        Trigger.TriggerState state = scheduler.getTriggerState(new TriggerKey(getScheduleId(program, programType,
-                                                                                            scheduleName)));
-        // Map trigger state to schedule state.
-        // This method is only interested in returning if the scheduler is
-        // Paused, Scheduled or NotFound.
-        switch (state) {
-          case NONE:
-            return ScheduleState.NOT_FOUND;
-          case PAUSED:
-            return ScheduleState.SUSPENDED;
-          default:
-            return ScheduleState.SCHEDULED;
-        }
-      } catch (SchedulerException e) {
-        throw Throwables.propagate(e);
-      }
+    switch (type) {
+      case STREAM:
+        return streamSizeScheduler.scheduleState(program, programType, scheduleName);
+      case TIME:
+        return timeScheduler.scheduleState(program, programType, scheduleName);
     }
+    throw new IllegalStateException("Unhandled type of schedule: " + type);
+  }
 
-    private void checkInitialized() {
-      Preconditions.checkNotNull(scheduler, "Scheduler not yet initialized");
+  private Store getStore() {
+    if (store == null) {
+      store = storeFactory.create();
     }
+    return store;
+  }
 
-    private String getScheduleId(Id.Program program, SchedulableProgramType programType, String scheduleName) {
-      return String.format("%s:%s", getJobKey(program, programType).getName(), scheduleName);
+  private ScheduleType getScheduleType(Id.Program program, SchedulableProgramType programType, String scheduleName) {
+    ApplicationSpecification appSpec = getStore().getApplication(program.getApplication());
+    if (appSpec == null) {
+      return null;
```

Comment:
I am a little worried about the null returns. Sometimes we throw an IllegalStateException, sometimes we return null - and then the caller just ignores the call? What kind of error handling is this? Is it expected that we simply ignore the call if the app, the program or the schedule is not known But throw an exception if the scheduler type is not known?